src/System.Runtime.Extensions/src/System/Environment.Win32.cs

```diff
@@ -2,60 +2,108 @@
+                }
+                else
                 {
-                    userDomainName = samName.Substring(0, index);
+                    builder.Length = 0;
                     return;
                 }
             }
-            domainNameLen = (uint)domainName.Capacity;
 
-            byte[] sid = new byte[1024];
-            int sidLen = sid.Length;
-            int peUse;
-            if (!Interop.Advapi32.LookupAccountNameW(null, UserName, sid, ref sidLen, domainName, ref domainNameLen, out peUse))
+            builder.Length = (int)size;
+        }
+
+        public static string UserDomainName
+        {
+            get
             {
-                throw new InvalidOperationException(Win32Marshal.GetExceptionForLastWin32Error().Message);
-            }
+                // See the comment in UserName
+                Span<char> initialBuffer = stackalloc char[40];
+                ValueStringBuilder builder = new ValueStringBuilder(initialBuffer);
+                GetUserName(ref builder);
+
+                var name = builder.AsSpan();
+                int index = name.IndexOf('\\');
+                if (index != -1)
+                {
+                    // In the form of DOMAIN\User, cut off \User and return
+                    return name.Slice(0, index).ToString();
+                }
+
+                // In theory we should never get use out of LookupAccountNameW as the above API should
+                // always return what we need. Can't find any clues in the historical sources, however.
+
+                // Domain names aren't typically long.
+                // https://support.microsoft.com/en-us/help/909264/naming-conventions-in-active-directory-for-computers-domains-sites-and
+                Span<char> initialDomainNameBuffer = stackalloc char[64];
+                ValueStringBuilder domainBuilder = new ValueStringBuilder(initialBuffer);
+                uint length = (uint)domainBuilder.Capacity;
 
-            userDomainName = domainName.ToString();
+                // This API will fail to return the domain name without a buffer for the SID.
+                // SIDs are never over 68 bytes long.
+                Span<byte> sid = stackalloc byte[68];
+                uint sidLength = 68;
+
+                while (!Interop.Advapi32.LookupAccountNameW(null, ref builder.GetPinnableReference(), ref MemoryMarshal.GetReference(sid),
+                    ref sidLength, ref domainBuilder.GetPinnableReference(), ref length, out _))
+                {
+                    int error = Marshal.GetLastWin32Error();
+
+                    // The docs don't call this out clearly, but experimenting shows that the error returned is the following.
+                    if (Marshal.GetLastWin32Error() != Interop.Errors.ERROR_INSUFFICIENT_BUFFER)
+                    {
+                        throw new InvalidOperationException(Win32Marshal.GetExceptionForWin32Error(error).Message);
```

Comment:
Why not just call `Interop.Kernel32.GetMessage(errorCode)` ?

Reply:
Good question- I copied the existing code. I'll change it if I can.

Reply:
Win32Marshal has a thin wrapper around it, I'll call that.